File: problems/dfs/on_tree/visible_tree_node/attempt.js
"use strict";

class Node {
	constructor(val, left = null, right = null) {
		this.val = val;
		this.left = left;
		this.right = right;
	}
}

function dfs(root, highest) {
	if (root === null) {
		return 0;
	}
	let total = 0;
	if (root.val >= highest) total++;

	total += dfs(root.left, Math.max(root.val, highest));
	total += dfs(root.right, Math.max(root.val, highest));
	return total;
}

function visibleTreeNode(root) {
	return dfs(root, Number.NEGATIVE_INFINITY);
}

function buildTree(nodes, f) {
	const val = nodes.next().value;
	if (val === "x") return null;
	const left = buildTree(nodes, f);
	const right = buildTree(nodes, f);
	return new Node(f(val), left, right);
}

function splitWords(s) {
	return s === "" ? [] : s.split(" ");
}

function* main() {
	const root = buildTree(splitWords(yield)[Symbol.iterator](), parseInt);
	const res = visibleTreeNode(root); 
	console.log(res);
}

class EOFError extends Error {}

{
    const gen = main();
    const next = (line) => gen.next(line).done && process.exit();
    let buf = "";
    next();
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (data) => {
        const lines = (buf + data).split("\n");
        buf = lines.pop();
        lines.forEach(next);
    });
    process.stdin.on("end", () => {
        buf && next(buf);
        gen.throw(new EOFError());
    });
}

File: problems/dfs/on_tree/visible_tree_node/attempt.py
class Node:
	def __init__(self, val, left=None, right=None):
		self.val = val
		self.left = left
		self.right = right

def visible_tree_node(root: Node) -> int:
	def dfs(root: Node, highest: int) -> int:
		if root is None:
			return 0
		res = 0
		if root.val >= highest:
			res += 1
		res += dfs(root.left, max(root.val, highest))
		res += dfs(root.right, max(root.val, highest))
		return res
	return dfs(root, float('-inf'))
# this function builds a tree from input; you don't have to modify it
# learn more about how trees are encoded in https://algo.monster/problems/serializing_tree
def build_tree(nodes, f):
	val = next(nodes)
	if val == "x":
		return None
	left = build_tree(nodes, f)
	right = build_tree(nodes, f)
	return Node(f(val), left, right)

if __name__ == "__main__":
	root = build_tree(iter(input().split()), int)
	res = visible_tree_node(root)
	print(res)

